### ISSUE TITLE
Device for sample_b in rff.functional

### ISSUE THREAD

**Issue by pulkitgo** [jmclong/random-fourier-features-pytorch#2]
**Describe the bug**
The random tensor returned by `sample_b` in `rff / functional.py` is not initialised on the device user might be using (since by default it would load on CPU). It can be corrected by including a device argument in the `GaussianEncoding()` initialization and subsequently in the `sample_b` function, so that `b` is  loaded on appropriate device.

```python
def sample_b(sigma: float, size: tuple) -> Tensor:
    r"""Matrix of size :attr:`size` sampled from from :math:`\mathcal{N}(0, \sigma^2)`
    Args:
        sigma (float): standard deviation
        size (tuple) size of the matrix sampled
    See :class:`~rff.layers.GaussianEncoding` for more details
    """
    return torch.randn(size) * sigma
```

**To Reproduce**
1. Initialise a random input tensor on device='cuda', 
2. Initialize a GaussianEncoding() layer and use it to process above input.

**Expected behavior**
Should run without any device mismatch errors for matrix multiplication.

**Config**
- Python 3.7.10
- PyTorch 1.8 + CUDA 11.0

**Comment by jmclong** [comment]
Hi, thank for the bug report.

I understand correctly, you are doing the following:
```python
import torch
import rff.layer
x = torch.ones(1,1)
x = x.to('cuda')
f  = rff.layers.GaussianEncoding(1, 1, 1)
f(x)
```

and you see the error
```
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking arugment for argument mat2 in method wrapper_mm)
```

I noticed that doing this with a linear layer from PyTorch produces the same results:


```python
import torch
import torch.nn as nn

x = torch.ones(1,1)
x = x.to('cuda')
f = nn.Linear(1,1)
f(x)
```

It should be ok if you send the model to cuda as well, which sends all parameters to the GPU:

```python
import torch
import rff.layer
x = torch.ones(1,1)
x = x.to('cuda')
f  = rff.layers.GaussianEncoding(1, 1, 1).to('cuda')
f(x)
```

Let me know if that helps!

**Comment by jmclong** [comment]
I'm going to close this

**Comment by sleepyeye** [comment]
Hello @jmclong 
It seems like `PositionalEncoding` layer still has similar issue. 
Here is the error message.

```python
    j = torch.arange(m)
    coeffs = 2 * np.pi * sigma ** (j / m)
    vp = coeffs * v[..., np.newaxis]
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    vp_cat = torch.cat((torch.cos(vp), torch.sin(vp)), dim=-1)
    return vp_cat.flatten(-2, -1)
```

This means that, if I do `encoder = rff.layers.PositionalEncoding(...).to('cuda')`, layer parameter are sent to the device.
But, the layer calls the function which depends on numpy function([here](https://github.com/jmclong/random-fourier-features-pytorch/blob/22ad7db8387c3df397e828e93091dc26ba4c2b9b/rff/functional.py#L74-L75)). 

I think the numpy function can be replaced with pytorch function.

**Comment by jmclong** [comment]
Hi @sleepyeye,

Thank you for reporting this issue. 

I think that` torch.unsqueeze(v, -1)` might resolve this issue. 

Can you try replacing  `v[..., np.newaxis]` with that ans see if it resolves this issue.

Thanks again.

Best,
Joshua

**Comment by jmclong** [comment]
This was not an issue in the `GaussianEncoding`, but is a real issue in `PositionalEncoding`

**Comment by jmclong** [comment]
@sleepyeye, sorry for really terrible delay, but this issue should be fixed now.

You should be able to get the fixed version from pip as version 0.2.

Thanks,
Joshua
